static/app/components/modals/inviteMembersModal/useInviteModal.tsx

```diff
@@ -0,0 +1,224 @@
+import {useCallback, useEffect, useMemo, useRef, useState} from 'react';
+
+import {
+  InviteRow,
+  NormalizedInvite,
+} from 'sentry/components/modals/inviteMembersModal/types';
+import {t} from 'sentry/locale';
+import {Member, Organization} from 'sentry/types';
+import {trackAnalytics} from 'sentry/utils/analytics';
+import {uniqueId} from 'sentry/utils/guid';
+import {useApiQuery} from 'sentry/utils/queryClient';
+import useApi from 'sentry/utils/useApi';
+
+interface Props {
+  organization: Organization;
+  initialData?: Partial<InviteRow>[];
+  source?: string;
+}
+function defaultInvite(): InviteRow {
+  return {
+    emails: new Set<string>(),
+    teams: new Set<string>(),
+    role: 'member',
+  };
+}
+
+function useLogInviteModalOpened({
+  organization,
+  sessionId,
+  source,
+}: {
+  organization: Organization;
+  sessionId: string;
+  source: string | undefined;
+}) {
+  useEffect(() => {
+    trackAnalytics('invite_modal.opened', {
+      organization,
+      modal_session: sessionId,
+      can_invite: organization.access?.includes('member:write'),
+      source,
+    });
+  }, [organization, sessionId, source]);
+}
+
+export default function useInviteModal({organization, initialData, source}: Props) {
+  const api = useApi();
+  const willInvite = organization.access?.includes('member:write');
+
+  /**
+   * Used for analytics tracking of the modals usage.
+   */
+  const sessionId = useRef(uniqueId());
+  useLogInviteModalOpened({organization, sessionId: sessionId.current, source});
+
+  const memberResult = useApiQuery<Member>(
+    [`/organizations/${organization.slug}/members/me/`],
+    {
+      staleTime: 0,
+    }
+  );
+
+  const [state, setState] = useState(() => {
+    return {
+      pendingInvites: initialData
+        ? initialData.map(initial => ({
+            ...defaultInvite(),
+            ...initial,
+          }))
+        : [defaultInvite()],
+      inviteStatus: {},
+      complete: false,
+      sendingInvites: false,
+    };
+  });
+
+  const invites = useMemo(() => {
+    return state.pendingInvites.reduce<NormalizedInvite[]>(
+      (acc, row) => [
+        ...acc,
+        ...[...row.emails].map(email => ({email, teams: row.teams, role: row.role})),
```

Comment:
map already returns a new array
```suggestion
        ...row.emails.map(email => ({email, teams: row.teams, role: row.role})),
```

Reply:
Minor, but there is also no need to spread the acc as it'll result in quadratic runtime. You can just return acc.concat(row.emails.map)